### PR TITLE
Feature/Env Variables

### PR DESCRIPTION
**Pull request by HenryHengZJ**
- Allow user to add variables to be used in custom tool function
- Allow sessionid, chatid to be used in custom tool function

### CODE REVIEW

File: packages/components/nodes/memory/ZepMemory/ZepMemory.ts [hunk 1]
```diff
@@ -196,6 +197,28 @@ class ZepMemoryExtended extends ZepMemory {
         }
         return super.clear()
     }
+
+    async getChatMessages(overrideSessionId = '', returnBaseMessages = false): Promise<IMessage[] | BaseMessage[]> {
```
Comment (vinodkiran) [inline packages/components/nodes/memory/ZepMemory/ZepMemory.ts:201]: Should this class implement MemoryMethods?

File: packages/components/src/Interface.ts [hunk 2]
```diff
@@ -195,3 +196,34 @@ export class VectorStoreRetriever {
...
+import { BufferMemory, BufferWindowMemory, ConversationSummaryMemory } from 'langchain/memory'
+
+export interface MemoryMethods {
+    getChatMessages(overrideSessionId?: string, returnBaseMessages?: boolean): Promise<IMessage[] | BaseMessage[]>
+    addChatMessages(msgArray: { text: string; type: MessageType }[], overrideSessionId?: string): Promise<void>
+    clearChatMessages(overrideSessionId?: string): Promise<void>
+    resumeMessages?(messages: IMessage[]): Promise<void>
+}
+
+export abstract class FlowiseMemory extends BufferMemory implements MemoryMethods {
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/components/src/Interface.ts:213]: make sense. So we can remove all these and retain the `MemoryMethods`, then implement this method for every other memory nodes

File: packages/components/nodes/memory/BufferMemory/BufferMemory.ts [hunk 3]
```diff
@@ -41,12 +42,49 @@ class BufferMemory_Memory implements INode {
...
-        return new BufferMemory({
+        return new BufferMemoryExtended({
             returnMessages: true,
             memoryKey,
             inputKey
         })
     }
 }
 
+class BufferMemoryExtended extends FlowiseMemory implements MemoryMethods {
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/components/nodes/memory/BufferMemory/BufferMemory.ts:53]: if you dont implment the methods, and when you have an error on the function, like `addChatMessages` to `addChatMesssages`, you dont get an error:
image

File: packages/components/src/Interface.ts [hunk 4]
```diff
@@ -195,3 +196,34 @@ export class VectorStoreRetriever {
...
+    abstract getChatMessages(overrideSessionId?: string, returnBaseMessages?: boolean): Promise<IMessage[] | BaseMessage[]>
+    abstract addChatMessages(msgArray: { text: string; type: MessageType }[], overrideSessionId?: string): Promise<void>
+    abstract clearChatMessages(overrideSessionId?: string): Promise<void>
+}
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/components/src/Interface.ts:218]: added `resumeMessages`

File: packages/server/src/utils/index.ts [hunk 5]
```diff
@@ -1050,3 +1052,29 @@ export const getAllValuesFromJson = (obj: any): any[] => {
     extractValues(obj)
     return values
 }
+
+export const replaceEnvVariables = async (question: string, appDataSource: DataSource): Promise<string> => {
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/server/src/utils/index.ts]: removed

File: packages/components/nodes/memory/MotorheadMemory/MotorheadMemory.ts [hunk 6]
```diff
@@ -134,12 +139,29 @@ interface MotorheadMemoryExtendedInput {
 class MotorheadMemoryExtended extends MotorheadMemory {
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/components/nodes/memory/MotorheadMemory/MotorheadMemory.ts]: implemented `MemoryMethods`

File: packages/components/nodes/memory/ZepMemory/ZepMemory.ts (hunk 1 above)
Comment (HenryHengZJ) [reply packages/components/nodes/memory/ZepMemory/ZepMemory.ts:201]: implemented `MemoryMethods`

File: packages/ui/src/views/variables/index.js [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+                                        startIcon={<IconPlus />}
+                                    >
+                                        Add Variable
+                                    </StyledButton>
+                                </ButtonGroup>
+                            </ButtonGroup>
+                        </Toolbar>
+                    </Box>
+                </Stack>
+                {variables.length <= 0 && (
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/ui/src/views/variables/index.js]: updated

**Review by vinodkiran — APPROVED** [review]

**Pull request merged**